sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala

```diff
@@ -43,6 +46,33 @@ private[sql] class DataSourceRegistration private[sql] (dataSourceManager: DataS
          | pythonExec: ${dataSource.dataSourceCls.pythonExec}
       """.stripMargin)
 
+    checkDataSourceExists(name)
+
     dataSourceManager.registerDataSource(name, dataSource)
   }
+
+  /**
+   * Checks if the specified data source exists.
+   *
+   * This method allows for user-defined data sources to be registered even if they
+   * have the same name as an existing data source in the registry. However, if the
+   * data source can be successfully loaded and is not a user-defined one, an error
+   * is thrown to prevent lookup errors with built-in or Scala/Java data sources.
+   */
+  private def checkDataSourceExists(name: String): Unit = {
+    // Allow re-registration of user-defined data sources.
+    if (dataSourceManager.dataSourceExists(name)) return
+
+    try {
+      DataSource.lookupDataSource(name, conf)
```

Comment:
I feel like we should probably scope this into runtime datasources (can be overwritten), and static datasources (cannot be overwritten). This change should probably be fine for now though.

Reply:
Maybe just call `SQLConf.get` for `conf`.

Reply:
Another concern is that when the name conflicts with statically registered sources..

Reply:
The rule I am considering is that:
- If the name of any user-defined data source conflicts with a built-in data source or an existing **Java/Scala** data source that can be loaded from the classpath, an exception will be thrown.
- Otherwise, data sources can be overridden.

For static Python data sources, should we allow them to be overridden?

Reply:
That makes sense in a way. But technically we can also overwrite Java/Scala datasources once we allow runtime registration, and my concern is that it might be somewhat inconsistent and difficult to understand how the overwrite works to the end users. One way is just to allow overwrite always, or disallow at all.

Reply:
So I was thinking that statically registered ones cannot be overwritten but not sure which way is better.

Reply:
I agree we should keep it consistent by blocking all statically registered data sources (we can change this based on user feedback in the future).
Currently, it appears we can't differentiate between a statically registered Python data source and a dynamic one. Perhaps we could add a flag in the dataSourceBuilders to indicate whether it's static or dynamic.

Reply:
I am fine with this for now but should probably make sure ^ change happen before Spark 4.0.

Reply:
Sounds good! Will add a TODO here.